phsb5321/aca.so
# Chore/code refactor

# Refactoring to implement basic separation of concerns

This PR implements the following changes:

- Separation of concerns between the `usecases` and `repository` modules.
- Creation of a `base_repository` module to abstract common database operations.
- Implementation of error abstractions to provide consistent error messages.
- Implementation of a `gremlin_connection` context manager to handle database connections.

This refactoring aims to improve the maintainability and scalability of the codebase by adhering to basic separation of concerns principles and reducing code duplication.

## Changes Made

To achieve the above goals, the following changes were made:

- Moved the Gremlin database connection to the `graph.py` module.
- Created the `base_repository.py` module to abstract common database operations.
- Updated the `usecases` modules to only contain business logic and to use the `base_repository` module for database operations.
- Implemented error abstractions to provide consistent error messages across the API.
- Created the `gremlin_connection` context manager to handle database connections and ensure they are properly closed.

## Testing

To test these changes, follow these steps:

1. Clone this repository and navigate to the root directory.
2. Install the required dependencies by running `poetry install`.
3. Start the Gremlin server by running `docker-compose up -d`.
4. Create a new virtual environment and activate it.
5. Start the API by running `uvicorn app.main:app`.
6. Use a tool like `curl`, `httpie`, or `Postman` to make requests to the API endpoints.
7. Verify that the responses match the expected results.
